doc/config.js.md
>**[Return to the Masterportal documentation](doc.md)**

[TOC]

# config.js

The `config.js` contains Masterportal configuration not directly related to UI or layers. For example, paths to other configuration files belong here. This file is usually placed next to the `index.html` and `config.json` files.

In the following, all configuration options are described. For all configuration options of type `object`, further nested options are linked and described in detail after the main table. You may also refer to **[this config.js example file](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/portal/basic/config.js)**.

|Name|Required|Type|Default|Description|Example|
|----|--------|----|-------|-----------|-------|
|layerConf|yes|String||Path to the **[services.json](services.json.md)** file containing all available WMS layers and WFS feature types. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/services-internet.json"`|
|namedProjections|yes|String[]||Definition of the usable coordinate systems. See **[syntax definition](http://proj4js.org/#named-projections)** for details..|`[["EPSG:25832", "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"]]`|
|restConf|yes|String||Path to the **[rest-services.json](rest-services.json.md)** file describing further services, e.g. print service, WPS, CSW. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/rest-services-internet.json"`|
|styleConf|yes|String||Path to the **[style.json](style.json.md)** file describing vector layer (WFS) styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/style.json"`|
|addons|no|String[]|`[]`|List of names for custom modules. The modules are to be placed in the folder `/addons/`, with their entry points being defined in the `addonsConf.json`.|`["myAddon1", "myAddon2"]`|
|alerting|no|**[alerting](#markdown-header-alerting)**|`{"category": "alert-info", "isDismissable": true, "isConfirmable": false, "position": "top-center", "fadeOut": null}`|Overrides the alert module's default values.|{fadeOut: 6000}|
|cameraParameter|no|**[cameraParameter](#markdown-header-cameraparameter)**||Initial camera parameter||
|cesiumParameter|no|**[cesiumParameter](#markdown-header-cesiumparameter)**||Cesium flags||
|cswId|no|String|`"3"`|Reference to a CSW interface used to retrieve layer information. The ID will be resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`"my CSW-ID"`|
|defaultToolId|no|String|`"gfi"`|The tool with the given ID will be active when no other tool is active.|"filter"|
|layerSelector|no|**[layerSelector](#markdown-header-layerselector)**||Module to configure interactions with the layertree and the map, executed on a defined event.||
|featureViaURL|no|**[featureViaURL](#markdown-header-featureviaurl)**||Optional configuration for the URL parameter `featureViaURL`. See **[urlParameter](urlParameter.md)** for details. Implemented for treeTypes *light* and *custom*.||
|footer|no|**[footer](#markdown-header-footer)**||If set, a footer is shown and configured with this object.||
|gfiWindow|no|String|`"detached"`|_Deprecated in the next major release. Please use the attribute "Portalconfig.menu.tool.gfi.desktopType" of the **[config.json](#config.json.md)** instead._ Display type and attribute information for all layer types. **attached**: the attribute information window is opened at click position **detached**: the attribute information window is opened at the top right of the map; a marker is set to the click position.|`"attached"`|
|ignoredKeys|no|String[]|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|List of attribute names to be ignored for attribute information lists of all layer types. Only used with "gfiAttributes": "showAll".|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|
|infoJson|no|String|`"info.json"`|Path to the `info.json` file containing additional information on snippets. The path is relative to the *index.html*.|`"info.json"`|
|inputMap|no|Object|`{}`|If this object is set, and its field `setMarker` is set to true, the Masterportal is configured as input element for data. In that case, each click sets a Map Marker and communicates the coordinates via **[remoteInterface](remoteInterface.md)** in the chosen coordinate reference system.|`{setMarker: true, targetProjection: "EPSG:4326", setCenter: false}`|
|inputMap.setCenter|no|Boolean|`false`|Center on a marker after producing it?|`setCenter: true`|
|inputMap.setMarker|no|Boolean|`false`|Flag to activate the 'setMarker' functionality.|`setMarker: true`|
|inputMap.targetProjection|no|String|`"EPSG:25832"`|The target coordinate reference system. Coordinates will be translated to it before being communicated via **[remoteInterface](remoteInterface.md)**.|`targetprojection: "EPSG:4326"`|
|mapMarker|no|**[mapMarker](#markdown-header-mapmarker)**||Overrides the map marker module's default values. Useful for 3D markers since OpenLayers's overlays can not be displayed in 3D mode. For this, the map marker has to be defined as vector layer.||
|metaDataCatalogueId|no|String|`"2"`|URL to the metadata catalog linked to in the layer information window. The ID is resolved to a service of the **[rest-services.json](rest-services.json.md)** file. Note: This attribute is only necessary, when no "show_doc_url" is configured in the metadata dataset in the **[services.json](services.json.md)**. The url can either be set globally (**[config.js](config.js.md)**) or layer-specific(**[services.json](services.json.md)**).|`"MetaDataCatalogueUrl"`|
|metadata|no|**[metadata](#markdown-header-metadata)**||Allows configuration of which metadata URLs are to be resolved via proxy.||
|mouseHover|no|**[mouseHover](#markdown-header-mousehover)**||Activates the MouseHover feature for vector layers, both WFS and GeoJSON. For per-layer configuration, see the **[config.json](config.json.md)**'s section *Themenconfig.Fachdaten.Layer*.|`true`|
|obliqueMap|no|Boolean|`false`|If set to `true`, an oblique map layer is created. An additional oblique layer must be defined.||
|portalConf|no|String|`"config.json"`|Path to the portal's `config.json` file. You may also enter a node; in that case the taken path is controlled by the urlParameter `config`.|Direct path: "../masterTree/config.json"; Node: "../../portal/master/". In the node scenario, a query parameter like `config=config.json` must exist in the URL.|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal is supposed to post messages to and receive messages from with the `postMessage` feature.|"http://localhost:8080"|
|proxyHost|no|String||Host name of a remote proxy with CORS configured to support the portal's domain, among others.|`"https://proxy.example.com"`|
|quickHelp|no|Object|`{}`|Activates the QuickHelp module. This displays a window containing help text for supported functions of the modules. Available for the layer tree (CustomTree), the search bar (Searchbar) and the routing tool (RoutingTool).||
|remoteInterface|no|**[remoteInterface](#markdown-header-remoteinterface)**||Optional remote interface configuration.||
|scaleLine|no|Boolean|`false`|Controls whether a scale line is displayed at the bottom right of the map.|`true`|
|simpleMap|no|Boolean|`false`|_Deprecated in the next major release. Please use the parameter `simpleMap` as part of the configuration of the `saveSelection` tool in the **[config.json](config.json.md)**._ Adds a SimpleMap URL to the `Save selection` dialogue. When calling this URL, the menu bar, layer tree, and map controls are deactivated. Not implemented for tree type *„light“*.|`false`|
|startingMap3D|bi|Boolean|`false`|Controls whether the map should start in 3D mode.||
|tree|no|**[tree](#tree)**||||
|uiStyle|no|String|`"default"`|Sets the control element layout. |`table`|
|wfsImgPath|no|String||Path to the folder holding images for the WFS styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/img/"`|
|wpsID|no|String|`""`|Reference to a WPS interface used in various modules. The ID is resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`""`|
|zoomToFeature|no|**[zoomToFeature](#markdown-header-zoomtofeature)**||_Deprecated in the next major release. Please use **[zoomTo](#markdown-header-zoomto)** instead._ Optional configuration of the URL query parameter `featureid`. For details, see **[urlParameter](urlParameter.md)**. ||
|zoomTo|no|**[zoomTo](#markdown-header-zoomto)**[]|Configuration for the URL query parameters `zoomToFeatureId` and `zoomToGeometry`.||
|layerInformation|no|**[layerInformation](#markdown-header-layerinformation)**||Configuration for the layerInformation window.||
|vuetify|no|String|undefined|Path to the optional instance of the vuetify UI library. e.g. portal or addon specific.|`addons/cosi/vuetify/index.js`|

***

## alerting

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|fetchBroadcastUrl|no|String|`false`|The alerting module will initially use a linked configuration file from this URL, if set.|
|localStorageDisplayedAlertsKey|no|String|`"displayedAlerts"`|Arbitrary key used to store information regarding the alerting module in the browser's local storage.|

***

## cameraParameter

Cesium Scene camera settings in 3D mode.
_Deprecated in the next major release. Please use **[cesiumParameter](#markdown-header-cesiumParameter)** instead._

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|
|altitude|no|Number||Camera's initial height in meters|

***

## cesiumParameter

Cesium Scene settings in 3D mode.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Scene.html?classFilter=scene)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|camera|no|**[camera](#markdown-header-cesiumParametercamera)**||Cesium Scene camera settings in 3D mode.|
|fog|no|**[fog](#markdown-header-cesiumParameterfog)**||Cesium Scene fog settings in 3D mode.|
|fxaa|no|Boolean|`true`|activates *fast approximate anti-aliasing*|
|globe|no|**[globe](#markdown-header-cesiumParameterglobe)**||Cesium Scene globe settings in 3D mode.|
|maximumScreenSpaceError|no|Number|`2.0`|Detail level in which terrain/raster tiles are fetched. 4/3 is the highest quality level.|
|tileCacheSize|no|Number|`100`|terrain/raster tile cache size|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    },
    "fog": {
        "enabled": true
    },
    "fxaa": true,
    "globe": {
        "enableLighting": true
    },
    "maximumScreenSpaceError": 2,
    "tileCacheSize": 20,
}
```
***

### cesiumParameter.camera

Cesium Scene camera settings in 3D mode.
The camera is defined by a position, orientation, and view frustum.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Camera.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|altitude|no|Number||Camera's initial height in meters|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    }
}
```
***

### cesiumParameter.fog

Cesium Scene fog settings in 3D mode.
Blends the atmosphere to geometry far from the camera for horizon views.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Fog.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|no|Boolean|`false`|True if fog is enabled.|

**Example**

```json
{
    "fog": {
        "enabled": true
    }
}
```
***

### cesiumParameter.globe

Cesium Scene globe settings in 3D mode.
The globe rendered in the scene, including its terrain and imagery layers.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Globe.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enableLighting|no|Boolean|`false`|Activates light effects on the map based on the sun's position.|

**Example**

```json
{
    "globe": {
        "enableLighting": true
    }
}
```
***

## footer

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|urls|no|**[urls](#markdown-header-footerurls)**[]||Array of URL configuration objects.|
|showVersion|no|Boolean|`false`|If `true`, the Masterportal version number is included in the footer.|
|footerInfo|no|**[footerInfo](#markdown-header-footerfooterInfo)**[]||Array of information configuration objects.|

***

### footer.urls
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|alias|no|String|`"Landesbetrieb Geoinformation und Vermessung"`|Link text for desktop applications|
|alias_mobil|no|String|`"LGV"`|Link text for mobile applications|
|bezeichnung|no|String|`"Kartographie und Gestaltung: "`|Link prefix|
|url|no|String||the URL the `alias` (or `alias_mobile`) text links to, e.g. `"https://example.com/"`. You may also link to mail creation with e.g. `"mailto:my_mail@example.com"`|
|toolModelId|no|String|`"SdpDownload"`|The id of a module to be opened on clicking the link. Do not include a URL for such cases.|

**Example:**

```json
{
    "footer": {
        "urls": [
            {
                "bezeichnung": "Cartography and design: ",
                "url": "https://geoinfo.hamburg.de/",
                "alias": "Landesbetrieb Geoniformation und Vermessung",
                "alias_mobil": "LGV"
            },
            {
                "bezeichnung": "",
                "url": "http://www.hamburg.de/bsu/timonline",
                "alias": "Map inconsistencies"
            },
            {
                "bezeichnung": "",
                "url": "",
                "alias": "SDP Download",
                "toolModelId": "sdpdownload"
            }
        ],
        "showVersion": true
    }
}
```

***

### footer.footerInfo
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|title|yes|String||Title of the information tab.|
|description|no|String||Text displayed under the title.|
|subtexts|no|**[subtexts](#markdown-header-footerfooterInfosubtexts)**[]||Array of subtext configuration objects.|

***

### footer.footerInfo.subtexts
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|subtitle|yes|String||Subtitle of the subtext.|
|text|no|String||Text displayed under the subtitle.|

**Beispiel:**

```json
{
    "footerInfo": [
        {
            "title": "Contact",
            "description": "Text under the titel",
            "subtexts": [
                {
                    "subtitle": "Postal address",
                    "text": "Max-Mustermann-Str. 1 <br> 12345 City <br> Germany"
                },
                {
                    "subtitle": "Phone and fax",
                    "text": "Tel: +49 (0) 1234 56789 <br> Fax: +49 (0) 1234 5678910"
                }
            ]
        },
        {
            "title": "Privacy",
            "subtexts": [
                {
                    "subtitle": "Subtitle",
                    "text": "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt..."
                }
            ]
        }
    ]
}
```
***

## mapMarker

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|pointStyleId|no|String|`"defaultMapMarkerPoint"`|StyleId to refer to a `style.json` point style. If not set, the `img/mapMarker.svg` is used.|
|polygonStyleId|no|String|`"defaultMapMarkerPolygon"`|StyleId to refer to a `style.json` polygon style.|

**Example:**

```json
{
    "mapMarker": {
        "pointStyleId": "customMapMarkerPoint",
        "polygonStyleId": "customMapMarkerPolygon"
    }
}
```

***

## mouseHover

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|minShift|no|Integer|`5`|Minimum mouse position movement required to render a new tooltip; in pixels.|
|numFeaturesToShow|no|Integer|`2`|Maximum amount of element information per tooltip; when exceeded, an information text informs the user of cut content.|
|infoText|no|String|`"(Further objects. Please zoom.)"`|Information text shown when `numFeaturesToShow` is exceeded.|

***

## portalLanguage

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|yes|Boolean|`true`|Controls whether a button to switch the portal's language is provided.|
|debug|no|Boolean|`false`|Controls whether debug information regarding translations is logged to the console.|
|languages|yes|Object|`{ de: "deutsch", en: "englisch" }`|Language abbreviations. Please mind that matching locale files must exist.|
|fallbackLanguage|no|String|`"de"`|Fallback language used if contents are not available in the currently selected language.|
|changeLanguageOnStartWhen|no|String[]|`["querystring", "localStorage", "navigator", "htmlTag"]`|Order of user language detection. See [i18next browser language detection documentation](https://github.com/i18next/i18next-browser-languageDetector) for details.|
|loadPath|no|String|`"/locales/{{lng}}/{{ns}}.json"`|Path to load language files from, or a function returning such a path: `function(lngs, namespaces) { return path; }`. `lng` and `ns` are read from the path, if given, as if from a static path. You may also provide a URL like `"https://localhost:9001/locales/{{lng}}/{{ns}}.json"`. See [i18next http backend documentation](https://github.com/i18next/i18next-http-backend) for details.|

**Example:**

```json
{
"portalLanguage": {
        "enabled": true,
        "debug": false,
        "languages": {
            "de": "deutsch",
            "en": "englisch"
        },
        "fallbackLanguage": "de",
        "changeLanguageOnStartWhen": ["querystring", "localStorage", "navigator", "htmlTag"],
        "loadPath": "/locales/{{lng}}/{{ns}}.json"
    }
}
```

***

## quickHelp

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|imgPath|no|String|`"/"`|An absolute or relative path to the folder containing the quick help images.|
|searchbarAllgemeines1|no|String|`"allgemein.png"`|First quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines2|no|String|`"allgemein_2.png"`|Second quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines3|no|String|`"allgemein_3.png"`|Third quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarFlurstueckssuche|no|String|`"allgemein_4.png"`|Quick help image regarding the Searchbar, belonging to the chapter "Parcel search". The image must exist in the `imgPath` folder.|
|aufbau1|no|String|`"themen.png"`|First quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|aufbau2|no|String|`"themen_2.png"`|Second quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|routingTool1|no|String|`"routing_1.png"`|First quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool2|no|String|`"routing_2.png"`|Second quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool3|no|String|`"routing_3.png"`|Third quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool4|no|String|`"routing_4.png"`|Fourth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool5|no|String|`"routing_5.png"`|Fifth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool6|no|String|`"routing_6.png"`|Sixth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool7|no|String|`"routing_7.png"`|Seventh quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool8|no|String|`"routing_8.png"`|Eighth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool9|no|String|`"routing_9.png"`|Nineth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool10|no|String|`"routing_10.png"`|Tenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool11|no|String|`"routing_11.png"`|Eleventh quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool12|no|String|`"routing_12.png"`|Twelfth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool13|no|String|`"routing_13.png"`|Thirteenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool14|no|String|`"routing_14.png"`|Fourteenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool15|no|String|`"routing_15.png"`|Fifteenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool16|no|String|`"routing_16.png"`|Sixteenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|

***

## remoteInterface
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal will post to and receive messages from with the `postMessage` feature.|

**Example:**

```json
{
    "remoteInterface": {
        "postMessageUrl": "http://localhost:8080"
    }
}
```
***

## tree

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|orderBy|no|String|`"OpenData"`|Category the layer tree is sorted by initially.|
|layerIDsToIgnore|no|Array||Array of `services.json` layer ids not to be shown in the layer tree.|
|layerIDsToStyle|no|**[layerIDsToStyle](#markdown-header-treelayeridstostyle)**[]||Special implementation for a HVV (Hamburg public transportation) service. Contains objects to request various styles of a layer id.|
|metaIDsToMerge|no|String[]||All layers found in the `services.json` regarding these meta IDs are merged to a single layer of the layer tree.|
|metaIDsToIgnore|no|String[]||All `services.json` layers listed will not be shown in the layer tree.|
|isFolderSelectable|no|Boolean|`true`|Globally sets whether a selection box is provided on folders that de-/activates all layers in it. An override per element exists, see **[config.json](config.json.md#Ordnerkonfiguration-Fachdaten)**.|

***

### tree.layerIDsToStyle

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|id|no|Sring||a `services.json` layer's id|
|styles|no|String/String[]||Style to be used as string; if multiple styles are to be used, they are listed in an array.|
|name|no|String/String[]||Name to be used as string; if multiple names are to be used, they are listed in an array.|
|legendUrl|no|String/String[]||Legend image URL to be used as string; if multiple legend images are to be used, their URLs are listed in an array.|

**Example:**

```json
{
    "tree": {

CHANGELOG.md
# Changelog Masterportal
 All important changes in this project are stored in this file.

[Semantic versioning](https://semver.org/spec/v2.0.0.html) is used.

## Known Issues

---

## Unreleased - in development
### __Breaking Changes__

### Added
- PortalFooter:
    - Possibility to display additional information in the footer
- Added DPI selection to print dialog (as advertised by mapfish print via "dpiSuggestions")

### Changed
- CompareFeatures: The values true and false are now translated.

### Deprecated

### Removed

### Fixed
- fix unix-related path issue in findWhereJs.js
- Issue #804: Legend window will not be cut off when width changes to a maximum.
- correct documentation regarding field names (serviceID -> serviceId)
- link in footer will now have white text for better readability.

---

## v2.23.0 - 2022-07-06
### Added
- FilterModule:
    - A new attribute clearAll for clearing all the filtered results after clicking button "Reset All"
- Api/highlightFeaturesByAttribute URL mechanism to query WFS data by parameters and configuration
- WfsSearch: A `multiSelect` mode has been added where multiple search results can be selected and viewn on the map.
- A new type of addons can be loaded: pure javascript-files with locale-files. See https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/doc/addOnsVue.md
- The elasticsearch now sends `layerIDsToIgnore` (as `id`) and `metaIDsToIgnore` (as `datasets.md_id`) to the service as well. These associated layers are filtered out of the response from Elastic.
- LayerSelector module:
    - Module to configure interactions with the layertree and the map, executed on a defined event.

### Changed
- Changed class for several title elements on page from span to h1 or h2 and standardized their styling.
- WMTS feature is now realized via masterportalAPI functions.
- SensorThingsAPI: The sensor layer has been moved from Backbone to Vue.
- Fix config.json.md regarding chapter depth.
- The following NPM packages are updated:
    - @masterportal/mpconfigparser: 1.2.0 to 1.3.0
    - @masterportal/masterportalapi: 2.3.0 to 2.4.0
  - Some `Radio.trigger` and `Radio.request` calls have been replaced by direct accesses to the `Vuex store`.

### Removed
- Deleted the following snippets: dropdown, exportButton, graphicalSelect, multiCheckbox.
- Removed the dependency `bootstrap-select`.

### Fixed
- Lines imported as KML can now be edited and modified with the Draw tool.
- Issue #663: Fixed tool Buffer Analysis where layer visibility updates would not be recognized in custom trees. Now, all visible layers at tool starting time are available within the tool.
- Issue #778: Fixed layer information display selection in metadata window that now always has names options in its select input.
- Issue #788: Legend not updating correctly.
- Download print Buttons will now be blue at first and grey once the button was clicked.
- Attached Gfi is styled correctly.
- Quickly clicking on a tab in the Elektro GFI will now not lead to a reload of the page.
- Tools without icon attribute in config.json use default icon from state now.
- MouseHover Module
  - Instead of getting all the layers from config.js, only the layers in portal will be loaded.
  - It does not depend on the gfi attributes to show the mouseHover tooltip.
  - Instead of text, the mouseHover tooltip will be shown as html content.
- The minimized ToolTemplate will now be blue and variables are used in css the template's css code.
- The transparency of a layer in the topic tree can now only be between 0 and 100 for increase or decrease.
- 3D-UrlParameters heading, tilt and altitude are now set.
- UrlParameter zoomtogeometry={number} is fixed.
- Fix error when defining a `styleID` for an OAF layer type.

---

## v2.22.2 - 2022-06-15
### Fixed
- Addons: Fixed tool Commuter flows.

---

## v2.22.1 - 2202-06-08
### Fixed
- The light grey color in Search and transparency are now dark
- Issue #790: Resolve import of isObject util failed was fixed.
- The Sensor Data GFI theme's layout is correct now.
- A console warning on creating an entities-layer was removed.

---

## v2.22.0 - 2022-06-01
### __Breaking Changes__
#### *Bootstrap upgrade*
The Bootstrap package was upgraded from 3.4.1 to 5.1.3. Masterportal developments with minor versions than Bootstrap 5.1.3 are no longer supported.

Please check if your own developments e.g. addons are affected and you have to make adaptions to the new bootstrap version.
Also the configuration of e.g. tools and controls in config.json has changed ("icon" instead of "glyphicon") and have to be updated. See [documentation](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/doc/config.json.md)

The upgrade included following necessary changes:

- Code-Refactoring to the new Bootstrap structure. Here new Bootstrap classes replaced old ones. For examples see the [Masterportal Refactoring](https://bitbucket.org/geowerkstatt-hamburg/masterportal/pull-requests/3406) or [Bootstrap documentation](https://getbootstrap.com/docs/5.0/migration/)
- Less is no longer supported, instead Sass has to be used for CSS. For examples see also [Masterportal Refactoring](https://bitbucket.org/geowerkstatt-hamburg/masterportal/pull-requests/3406)
- As icon set [Bootstrap Icons](https://icons.getbootstrap.com/) replaced glyphicons. Also, the configuration in config.json considering elements like tools and controls have changed. The old attribute "glypicon" has changed to "icon" and as values the new bootstrap-icons have to be used.

#### *Change in WFS filter parameter*
Issue #764: Using parameters in WfsSearch as defined in [Filter Encoding Implementation Specification](https://portal.ogc.org/files/?artifact_id=8340); the default name for the like filter escape character is now `escapeChar` rather than `escape`. This requires a change of configuration for services deviating from the standard implementation, see docs.

### Added
- The following NPM packages are added:

portal/mobility-data/config.js
const Config = {
    addons: ["storyTellingTool"],
    ignoredKeys: [
        "BOUNDEDBY",
        "SHAPE",
        "SHAPE_LENGTH",
        "SHAPE_AREA",
        "OBJECTID",
        "GLOBALID",
        "GEOMETRY",
        "SHP",
        "SHP_AREA",
        "SHP_LENGTH",
        "GEOM"
    ],
    wfsImgPath: "https://geodienste.hamburg.de/lgv-config/img/",
    vuetify: "addons/storyTellingTool/vuetify",
    metadata: {
        useProxy: ["https://metaver.de/csw"]
    },
    tree: {
        orderBy: "opendata",
        saveSelection: true,

        layerIDsToIgnore: [
            "1912",
            "1913",
            "1914",
            "1915",
            "1916",
            "1917", // UESG
            "2298", // Straßenbaumkataster cache grau
            "1791", // nachträgliche Bodenrichtwerte lagetypisch 1964
            "8713" // Layer Schulinfo
        ],
        layerIDsToStyle: [
            {
                id: "1933",
                styles: "geofox_stations",
                name: "Haltestellen",
                legendURL:
                    "https://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bus.png"
            },
            {
                id: "1935",
                styles: [
                    "geofox_Faehre",
                    "geofox-bahn",
                    "geofox-bus",
                    "geofox_BusName"
                ],
                name: [
                    "Fährverbindungen",
                    "Bahnlinien",
                    "Buslinien",
                    "Busliniennummern"
                ],
                legendURL: [
                    "https://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-faehre.png",
                    "https://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bahn.png",
                    "https://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bus.png",
                    "https://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bus.png"
                ]
            }
        ],
        metaIDsToMerge: [
            "4AC1B569-65AA-4FAE-A5FC-E477DFE5D303", // Großraum- und Schwertransport-Routen in Hamburg
            "3EE8938B-FF9E-467B-AAA2-8534BB505580", // Bauschutzbereich § 12 LuftVG Hamburg
            "F691CFB0-D38F-4308-B12F-1671166FF181", // Flurstücke gelb
            "FE4DAF57-2AF6-434D-85E3-220A20B8C0F1" // Flurstücke schwarz
        ],
        metaIDsToIgnore: [
            "09DE39AB-A965-45F4-B8F9-0C339A45B154", // MRH Fachdaten
            "51656D3F-E801-497C-952C-4F1F605843DD", // MRH Metrokarte
            "AD579C62-0471-4FA5-8C9A-38B3DCB5B2CB", // MRH Übersichtskarte-blau
            "14E3AFAE-99BE-4F1D-A3A6-6A68A1CDAC7B", // MRH Übersichtskarte-grün
            "56110E55-72C7-41F2-9F92-1C598E4E0A02", // Digitale Karte Metropolregion
            "88A22736-FE87-46F7-8A38-84F9E0E945F7", // TN für Olympia
            "DDB01922-D7B5-4323-9DDF-B68A42C559E6", // Olympiastandorte
            "AA06AD76-6110-4718-89E1-F1EDDA1DF4CF", // Regionales Raumordnungsprogramm Stade+Rotenburg
            "1C8086F7-059F-4ACF-96C5-7AFEB8F8B751", // Fachdaten der Metropolregion
            "A46086BA-4A4C-48A4-AC1D-9735DDB4FDDE", // Denkmalkartierung FIS
            "DB433BD1-1640-4FBC-A879-72402BD5CFDB", // Bodenrichtwertzonen Hamburg
            "6A0D8B9D-1BBD-441B-BA5C-6159EE41EE71", // Bodenrichtwerte für Hamburg
            "3233E124-E576-4B5D-978E-164720C4E75F", // MRH Große Verkehrsprojekte
            "24513F73-D928-450C-A334-E30037945729", // 3D Straßenbaumkataster Hamburg
            "7595A206-F07E-470D-A6C1-2F74F0B0C64E" // 3D Hamburger Hauptkirchen
        ]
    },
    scaleLine: true,
    quickHelp: {
        imgPath: "https://geodienste.hamburg.de/lgv-config/img/"
    },
    allowParametricURL: true,
    view: {
        center: [565874, 5934140]
    },
    namedProjections: [
        // GK DHDN
        [
            "EPSG:31467",
            "+title=Bessel/Gauß-Krüger 3 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=bessel +datum=potsdam +units=m +no_defs"
        ],
        // ETRS89 UTM
        [
            "EPSG:25832",
            "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"
        ],
        // LS 320: zusammen mit Detlef Koch eingepflegt und geprüft
        [
            "EPSG:8395",
            "+title=ETRS89/Gauß-Krüger 3 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=GRS80 +datum=GRS80 +units=m +no_defs"
        ],
        // WGS84
        [
            "EPSG:4326",
            "+title=WGS 84 (long/lat) +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"
        ]
    ],
    layerConf: "https://geodienste.hamburg.de/services-internet.json",
    restConf:
        "https://geodienste.hamburg.de/lgv-config/rest-services-internet.json",
    uiStyle: "default",
    styleConf: "https://raw.githubusercontent.com/herzogrh/faircare-verkehr/main/assets/data/style.json",
    storyConf: "./assets/story-3d.json",
    isMenubarVisible: true,
    gemarkungen: "https://geodienste.hamburg.de/lgv-config/gemarkung.json",
    obliqueMap: true,
    cesiumParameter: {
        tileCacheSize: 20,
        enableLighting: true,
        fog: {
            enabled: true,
            density: 0.0002,
            screenSpaceErrorFactor: 2.0
        },
        maximumScreenSpaceError: 2,
        fxaa: true
    },
    startingMap3D: false,
    portalLanguage: {
        enabled: true,
        debug: false,
        languages: {
            de: "deutsch",
            en: "englisch"
        },
        fallbackLanguage: "de",
        changeLanguageOnStartWhen: ["querystring", "localStorage", "htmlTag"]
    }
};

// conditional export to make config readable by e2e tests
if (typeof module !== "undefined") {
    module.exports = Config;
}
